seismicpro/src/gather/utils/general_utils.py

```diff
@@ -125,3 +125,59 @@ def mute_gather(gather_data, muting_times, samples, fill_value):
     mask = mask.reshape(-1)
     gather_data[~mask] = fill_value
     return gather_data.reshape(data_shape)
+
+
+@njit(parallel=True)
+def apply_agc(data, factor=1, window_size=125, mode='abs'):
+    """ TODO """
+    n_traces, trace_len = data.shape
+    win_left, win_right = window_size // 2, window_size - window_size // 2
+    # start is the first trace index that fits the full window, end is the last.
+    # AGC coefficients before start and after end are extrapolated.
+    start, end = win_left, trace_len - win_right
+
+    for i in prange(n_traces):  # pylint: disable=not-an-iterable
+        trace = data[i]
+        trace = np.power(trace, 2) if mode=='rms' else np.abs(trace)
+
+        amplitudes_cumsum = np.cumsum(trace)
+        nonzero_counts_cumsum = np.cumsum(trace!=0)
+
+        coefs = np.empty_like(trace)
+        coefs[start:end] = ((nonzero_counts_cumsum[:-window_size] - nonzero_counts_cumsum[window_size:])
+                            / (amplitudes_cumsum[:-window_size] - amplitudes_cumsum[window_size:] + 1e-15))
+        # Extrapolate AGC coefs for trace indices that don't fit the full window
+        coefs[:start] = coefs[start]
+        coefs[end:] = coefs[end-1]
+
+        coefs = np.sqrt(coefs) * factor if mode=='rms' else coefs * factor
+        data[i] *= coefs
+    return data
+
+
+@njit(parallel=True)
+def calculate_sdc_coefficient(v_pow, velocities, t_pow, times):
+    """ TODO """
+    sdc_coefficient = velocities**v_pow * times**t_pow
+    # Scale sdc_coefficient to be 1 at maximum time
+    sdc_coefficient /= sdc_coefficient[-1]
+    return sdc_coefficient
+
+
+@njit(parallel=True)
+def apply_sdc(data, v_pow, velocities, t_pow, times):
+    """ TODO """
+    n_traces, _ = data.shape
+    sdc_coefficient = calculate_sdc_coefficient(v_pow, velocities, t_pow, times)
+    for i in prange(n_traces):  # pylint: disable=not-an-iterable
+        data[i] *= sdc_coefficient
+    return data
+
+@njit(parallel=True)
+def undo_sdc(data, v_pow, velocities, t_pow, times):
+    """ TODO """
+    n_traces, _ = data.shape
+    sdc_coefficient = calculate_sdc_coefficient(v_pow, velocities, t_pow, times)
+    for i in prange(n_traces):  # pylint: disable=not-an-iterable
```

Comment:
```suggestion
    for i in prange(len(data)):  # pylint: disable=not-an-iterable
```
As well as in another func

Reply:
Done